src/main/java/com/google/devtools/build/lib/remote/merkletree/DirectoryTreeBuilder.java

```diff
@@ -262,10 +262,11 @@ private static <T> int build(
       PathFragment path = e.getKey();
       T input = e.getValue();
 
-      if (scrubber != null
-          && input instanceof ActionInput
-          && scrubber.shouldOmitInput((ActionInput) input)) {
-        continue;
+      if (scrubber != null && input instanceof ActionInput) {
+        String inputPath = input.equals(VirtualActionInput.EMPTY_MARKER) ? path.getPathString() : ((ActionInput)input).getExecPathString();
+        if (scrubber.shouldOmitInput(inputPath)) {
+          continue;
+        }
```

Comment:
Actually - I am now convinced that we should **always** scrub by `PathFragment`, not by`ActionInput.getExecPath()`.

The reason is runfile trees, which are the only situation in which `path != actionInput.getExecPath()`: the `path` refers to the location in the runfiles tree (e.g. `bazel-out/xxx/bin/exe.runfiles/foo/bar`) while the `actionInput` refers to the original artifact that will be staged there (e.g. `bazel-out/yyy/bin/foo/bar`). If we do the matching on the `actionInput`, a rule to scrub `bazel-out/yyy` won't apply because the `path` has a different prefix (`bazel-out/xxx`). This is basically the same bug you're seeing for empty files, but it applies equally to the non-empty ones.

In light of this, could you instead:

1. Change `shouldOmitInput` to take a `PathFragment` instead of an `ActionInput`
2. Amend this logic to always match on the `path` (no special case required for empty markers)
3. Amend the tests that are passing an `ActionInput` to `shouldOmitInputs` to pass in a `PathFragment` instead

Reply:
Makes sense. Will do this week. Thanks @tjgq !

Reply:
@oliviernotteghem are you still working on this? It would be great if it could make it into 7.1.0 (the first RC will be out today-ish, but we still have a few days to cherry-pick it)

Reply:
I will go ahead and do this myself, because it makes a followup change easier for me.